libs/checkpoint-duckdb/langgraph/store/duckdb/aio.py

```diff
@@ -0,0 +1,202 @@
+import asyncio
+import logging
+from contextlib import asynccontextmanager
+from typing import (
+    AsyncIterator,
+    Iterable,
+    Sequence,
+    cast,
+)
+
+import duckdb
+from langgraph.store.base import GetOp, ListNamespacesOp, Op, PutOp, Result, SearchOp
+from langgraph.store.base.batch import AsyncBatchedBaseStore
+from langgraph.store.duckdb.base import (
+    BaseDuckDBStore,
+    _convert_ns,
+    _group_ops,
+    _row_to_item,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncDuckDBStore(AsyncBatchedBaseStore, BaseDuckDBStore):
+    def __init__(
+        self,
+        conn: duckdb.DuckDBPyConnection,
+    ) -> None:
+        super().__init__()
+        self.conn = conn
+        self.loop = asyncio.get_running_loop()
+        self.is_setup = False
+
+    async def abatch(self, ops: Iterable[Op]) -> list[Result]:
+        await self.setup()
+        grouped_ops, num_ops = _group_ops(ops)
+        results: list[Result] = [None] * num_ops
+
+        tasks = []
+
+        if GetOp in grouped_ops:
+            tasks.append(
+                self._batch_get_ops(
+                    cast(Sequence[tuple[int, GetOp]], grouped_ops[GetOp]), results
+                )
+            )
+
+        if PutOp in grouped_ops:
```

Comment:
Are we worried at about race conditions? 

We have write operations mixed with read operations?

Reply:
mm, that's a good question. don't think we're currently calling them together anywhere, but it's probably good to avoid that -- i can add a warning?

Reply:
Is this exposed to end users?

Reply:
yes but we're not showing that anywhere in the docs. i.e. the APIs that are actually meant to be used are `.put()` / `.search()` / `.get`, which just call `.batch()` under the hood